src/io/asyncsocket.c

```diff
@@ -31,124 +31,150 @@ static void free_on_close_cb(uv_handle_t *handle) {
 static void on_read(uv_stream_t *handle, ssize_t nread, const uv_buf_t *buf) {
     ReadInfo         *ri  = (ReadInfo *)handle->data;
     MVMThreadContext *tc  = ri->tc;
-    MVMObject        *arr = MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTArray);
     MVMAsyncTask     *t   = MVM_io_eventloop_get_active_work(tc, ri->work_idx);
-    MVM_repr_push_o(tc, arr, t->body.schedulee);
-    if (nread >= 0) {
-        MVMROOT2(tc, t, arr, {
-            MVMArray *res_buf;
 
+    if (nread >= 0) {
+        MVMROOT3(tc, t, ri->handle, ri->buf_type, {
+            MVMObject *arr = MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTArray);
+            MVM_repr_push_o(tc, arr, t->body.schedulee);
             /* Push the sequence number. */
-            MVMObject *seq_boxed = MVM_repr_box_int(tc,
-                tc->instance->boot_types.BOOTInt, ri->seq_number++);
-            MVM_repr_push_o(tc, arr, seq_boxed);
-
-            /* Produce a buffer and push it. */
-            res_buf      = (MVMArray *)MVM_repr_alloc_init(tc, ri->buf_type);
-            res_buf->body.slots.i8 = (MVMint8 *)buf->base;
-            res_buf->body.start    = 0;
-            res_buf->body.ssize    = buf->len;
-            res_buf->body.elems    = nread;
-            MVM_repr_push_o(tc, arr, (MVMObject *)res_buf);
-
+            MVMROOT(tc, arr, {
+                MVMObject *seq_boxed = MVM_repr_box_int(tc,
+                    tc->instance->boot_types.BOOTInt, ri->seq_number++);
+                MVM_repr_push_o(tc, arr, seq_boxed);
+                /* Produce a buffer and push it. */
+                MVMROOT(tc, seq_boxed, {
```

Comment:
since seq_boxed doesn't get used in the code after this point, and it's already pushed into the array which makes it reachable, this MVMROOT can be dropped